working/2509-private-named-parameters/feature-specification.md

```diff
@@ -0,0 +1,382 @@
+
+This is a highly desired feature. The most-voted open issue in the language repo
+is for [data classes][]. That concept encompasses a few features, but if you
+look at the comments, many users are specifically requesting primary
+constructors. The issue for [primary constructors][primary ctors issue]
+specifically is the #11 upvoted request.
+
+[data classes]: https://github.com/dart-lang/language/issues/314
+[primary ctors issue]: https://github.com/dart-lang/language/issues/2364
+
+With that feature, users will encounter the limitation around private named
+parameters even more often. In-header primary constructors exacerbate this issue
+because an in-header primary constructor *can't* have an initializer list. Users
+will be forced to either make the field public, the parameter positional, or not
+use a primary constructor at all.
+
+We don't want to put users in a position where the code that's most pleasant to
+write requires them to sacrifice semantic properties they want like
+encapsulation.
+
+## Proposal
+
+The basic idea is simple. We let users use a private name in a named parameter.
+The compiler removes the `_` from the argument name but keeps it for the
+corresponding initialized or declared field. In other words, we do exactly what
+users are doing by hand when they write an initializer like:
+
+```dart
+class House {
+  int? _windows;
+  int? _bedrooms;
+  int? _swimmingPools;
+
+  House({
+    int? windows,
+    int? bedrooms,
+    int? swimmingPools,
+  })  : _windows = windows,
+        _bedrooms = bedrooms,
+        _swimmingPools = swimmingPools;
+}
+```
+
+With this proposal, the above class can be written with identical semantics like
+so:
+
+```dart
+class House {
+  int? _windows;
+  int? _bedrooms;
+  int? _swimmingPools;
+
+  House({this._windows, this._bedrooms, this._swimmingPools});
+}
+```
+
+When this proposal is combined with primary constructors, they can write:
+
+```dart
+class House({
+  int? _windows,
+  int? _bedrooms,
+  int? _swimmingPools,
+});
+```
+
+## Static semantics
+
+An identifier is a **private name** if it starts with an underscore (`_`),
+otherwise it's a **public name**.
+
+A private name may have a **corresponding public name**. If the characters of
+the identifier with the leading underscore removed form a valid identifier and a
+public name, then that is the private name's corresponding public name. *For
+example, the corresponding public name of `_foo` is `foo`.* If removing the
+underscore does not leave something which is is a valid identifier *(as in `_`
+or `_2x`)* or leaves another private name *(as in `__x`)*, then the private name
+has no corresponding public name.
+
+Given an initializing formal or field parameter (for a primary constructor) with
+private name *p* in constructor C:
+
+*   If *p* has no corresponding public name *n*, then compile-time error. *You
```

Comment:
Only if it's a named parameter?
Or do we not allow privately named positional parameters. (It's consistent, but more breaking, you can no longer do `foo(int _id, {String? id})` which you technically could before.

Reply:
Oops, this was supposed to be restricted to only affect the behavior of named parameters. Positional parameters should behave as they currently do where they can be private or not. Fixed.

Reply:
I'd actually *prefer* if positional parameters got the same treatment, just to make it explicit that
```dart
String value = "Something else";
class Box {
   final int _value;
   /// Creates a box with the given [value].
   C(this._value);
}
```
means that [value] definitely denotes the parameter.

We can make hacks in DartDoc to show public identifiers instead of private ones, so you write `the given [_value]` and it's shown as "The given <ul>value</ul>", but it's bound to cause headaches. Like here, where `value` actually denotes something else in the language.